src/Query/Group.ts

```diff
@@ -151,13 +162,44 @@ export class Group {
         return [Group.escapeMarkdownCharacters(filename)];
     }
 
-    private static groupByRoot(task: Task): string[] {
-        const path = task.path.replace(/\\/g, '/');
-        const separatorIndex = path.indexOf('/');
-        if (separatorIndex == -1) {
-            return ['/'];
+    private static groupByFn(task: Task, arg?: GroupingArg): string[] {
+        const paramsArgs: [string, any][] = [
+            ['description', task.description],
+            ['done', task.doneDate],
+            ['due', task.dueDate],
+            ['filename', task.filename],
+            ['happens', new HappensDateField().earliestDate(task)],
+            ['header', task.precedingHeader],
+            ['markdown', task.originalMarkdown],
+            ['path', task.path.replace('.md', '')],
+            ['priority', task.priority],
+            ['recurrence', task.recurrence],
+            ['root', Group.root(task)],
+            ['scheduled', task.scheduledDate],
+            ['start', task.startDate],
+            ['status', task.status],
+            ['t', task],
```

Comment:
Will this expose the `Task` object?

Reply:
Yes, as a variable `t` in the `group by fn` body. I also include it as a `task` var but it seemed useful to have a shorter name too 

```
group by fn t.status ===  'done' : 'Done' : 'Todo'
```

Reply:
I ask because I care a lot about avoiding breaking changes for users - not requiring them to rewrite tasks blocks if at all possible.

In #1249 I tried to explore whether making `Task` objects visible to users - with all its public data - would tie my hands and prevent me from doing future refactorings.

For that reason I started implementing an abstraction for dates, but didn't get very far as a lot of other stuff has cropped up.

I am therefore a little nervous about this PR's exposing of raw dates in this PR, and what restrictions it would mean on future improvements to code maintainability.

And considerably more nervous at the idea of exposing the entire `Task` class.

I am prepared to be persuaded away from that view, if someone can show me ways of exposing the inner details now, and not hampering future development... I just personally can't see a way right now.

The background to all this is that there are a load of missing abstractions that make addition of new features and comprehension of the code harder than I would like. And so I am trying to juggle improving the design with user support and PRs... (It's a nice position to be in, of course...)

Reply:
Oh I see, that all makes total sense. Thanks for explaining that to me.

Hmm, what if instead of my suggestion here https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1421#discussion_r1055916426 then, we went kinda the opposite direction and only exposed string values, no task/t, dates or complex objects.

```typescript
const paramsArgs: [string, any][] = [
    ['description', task.description],
    ['done', task.doneDate?.format('YYYY-MM-DD')],
    ['due', task.dueDate?.format('YYYY-MM-DD')],
    ['filename', task.filename],
    ['happens', new HappensDateField().earliestDate(task)?.format('YYYY-MM-DD')],
    ['heading', task.precedingHeader],
    ['markdown', task.originalMarkdown],
    ['path', task.path.replace('.md', '')],
    ['priority', task.priority],
    ['root', Group.root(task)],
    ['scheduled', task.scheduledDate?.format('YYYY-MM-DD')],
    ['start', task.startDate?.format('YYYY-MM-DD')],
    ['status', task.status.toLowerCase()],
    ['tags', task.tags],
    ['urgency', task.urgency],
];
```

Reply:
Thank you for being so understanding.

Your reply game me another idea. Refactoring the whole of tasks to use a new date class will take a while.

For info, I just pushed [refactor-add-TaskDate-class branch](https://github.com/obsidian-tasks-group/obsidian-tasks/commits/refactor-add-TaskDate-class) which is 2 months old and far, far for complete.

After reading you reply, I started to wonder - what if the dates in this PR were exposed as `TaskDate` objects, since that is the direction the code will be going in at some point.

I would like someone familiar with idiomatic TypeScript (I'm not) to review TaskDate and say if anything needs to be improved, but the class on its own could be made suitable for release in not very long. (I know it needs JSDocs).

So the diff from your above would look something like:

```diff
+ ['start', task.startDate?.format('YYYY-MM-DD')]
- ['start', new TaskDate(task.startDate)]
```

We could add methods to TaskDate based on what we learn from your work on this PR...

Does that make sense? What do you think?

Reply:
This makes total sense, yeah. FWIW I don't have much preference for which fields are exposed or in which format, etc. as long as I can (selfishly) still solve my original little problem lol. 😇 

Funny, I was even tempted to write a little date helper for my example (instead of repeating `?.format('YYYY-MM-DD')`) but wanted to stay focused. 

Wish I could be of more help re: idiomatic TypeScript and TaskDate, but I've only recently started using TypeScript myself. 🖤

But please let me know what else I can do to help with this PR anytime.

Reply:
> But please let me know what else I can do to help with this PR anytime.

I think the simplest thing is for you to continue with the plugin basically as is, just responding the code review feedback.
If you leave it returning `task.start` etc, then I can add the `TaskDate` class and hook it up after this PR is merged.
That way, you could focus on what you have already done, which is such a massive step forward.